Mappings for Getty and MWDL

marktriggs opened the pull request:
Here's a first attempt at mappings for Getty and MWDL.  Some general remarks:
- There's a lot of semicolon delimited fields, so I'm doing a lot of `.map { |v| v.value.split(';').map(&:strip) }.flatten` to deal with those.
  
  Does that seem like a sign that I should really be using an enrichment?  Or would it be worth defining a helper (on ValueArray, or a MappingTool?) for this?
- I've marked the places that I think might need enrichments with `TODO` entries.  Feedback on whether I've got that right would be appreciated :)
- The rule for working out the MWDL data provider is tricky, as it's conditional on the value of two different fields, and may be `nil` if neither field matches.  Originally I just had:
  
  ```
  intermediateProvider :class ... do
    label build_intermediate_provider
  end
  ```
  
  but when `build_intermediate_provider` would return `nil`, my JSON-LD output would have an `intermediateProvider` entry with an `@id` and `@type`, but no `name`, and that didn't seem right.
  
  I got around that using the `.map` call you can see in the code, but I wasn't sure if there was a better way to do this.

marktriggs committed:
Introduce mappings for Getty and MWDL

anarchivist commented:
>   * There's a lot of semicolon delimited fields, so I'm doing a lot of `.map { |v| v.value.split(';').map(&:strip) }.flatten` to deal with those.
>
>    Does that seem like a sign that I should really be using an enrichment?  Or would it be worth defining a helper (on ValueArray, or a MappingTool?) for this?

That sounds worthy of an enrichment to me, and I think [`Krikri::Enrichments::SplitAtDelimiter`](https://github.com/dpla/KriKri/blob/develop/lib/krikri/enrichments/split_at_delimiter.rb) probably will cover this need for the most part.  

>   * I've marked the places that I think might need enrichments with `TODO` entries.  Feedback on whether I've got that right would be appreciated :)

Again, these at an initial glance look like they might be enrichments. 

>  * The rule for working out the MWDL data provider is tricky, as it's conditional on the value of two different fields, and may be `nil` if neither field matches.  Originally I just had:
    ```
    intermediateProvider :class ... do
      label build_intermediate_provider
    end
    ```
    but when `build_intermediate_provider` would return `nil`, my JSON-LD output would have an `intermediateProvider` entry with an `@id` and `@type`, but no `name`, and that didn't seem right.
>
>    I got around that using the `.map` call you can see in the code, but I wasn't sure if there was a better way to do this.

You might want to see how it was implemented in the [University of Florida mapping](https://github.com/dpla/heidrun-mappings/blob/master/heidrun/ufl_marc.rb) - this may, or may not help.

anarchivist commented on heidrun/getty.rb:
```diff
@@ -0,0 +1,134 @@
+  end
+end
+
+lookup_language = lambda do |i|
```
Can you describe why this is necessary if `LANGUAGE_MAP` has only one key?

marktriggs replied:
Ah, just me being silly.  The language code will either be an iso639-2b code, or the string 'multiple languages', and in the latter case we just want to emit 'mul'.  I'd used a hash lookup with default to allow us to override multiple special values, but it's not really necessary.  I've switched it to a conditional to make this clearer.

anarchivist replied:
Ah, so for some reason it's not providing us with `mul`?

marktriggs replied:
Yep, that's right.  Strange...

no-reply replied:
I think we should be fine to just map the language string directly into `providedLabel`. Both cases here should be handled by [LanguageToLexvo](https://github.com/dpla/KriKri/blob/develop/lib/krikri/enrichments/language_to_lexvo.rb). For the 639-2 codes, those should all be present in 639-3. `"multiple languages"` is likewise present as a label, and will get treated as the language label and linked to `lexvo639_3:mul`, as in the same way a real language's name would.

(See `term :mul` at https://raw.githubusercontent.com/dpla/dpla_map/map-4.0/lib/rdf/iso_639_3.rb)

anarchivist commented on heidrun/getty.rb:
```diff
@@ -0,0 +1,134 @@
+                        Krikri::PrimoParser.display('lds03'))
+           .map { |v| v.value.split(';').map(&:strip) }.flatten
+
+    # TODO: The original (python) getty mapping seemed to use lds34 and lds43
```
cc @guegueng

guegueng replied:
This is for collection? 34 & 43 are correct. Thanks for catching that. I actually made the change in the spreadsheet, but then it got reverted by accident (looking at the revision history on the spreadsheet)

anarchivist commented on heidrun/getty.rb:
```diff
@@ -0,0 +1,134 @@
+      providedLabel creator
+    end
+
+    # TODO: Enrichment needed here to remove trailing punctuation and set
```
We should have an enrichment for the former already, at least.

markbreedlove replied:
I'm not sure what it would require to break out begin and end dates from their 'creationdate' field, but I'm guessing that it isn't necessary.  I assume that they can both be set to the same value (it being a single creation date), maybe even by our existing timespan-splitting enrichment (`timespan_split.rb`).  There are at least three providers for whom we do not assign any begin or end date in the mapping:  SCDL, DLG, and UW.

no-reply replied:
:+1: to @markbreedlove's comments. I think it would be fine to remove this comment this mapping to `providedLabel` seems :ok_hand: :100: to me.

marktriggs commented:
Cool, looking at the new Florida mapping, it appears I'm on the right track with how I handled intermediateProvider, so I'll keep it as is.

For the places in need of enrichments, is the best approach for me to add comments to the appropriate issues in Redmine?  Or is there some preferred way of tracking which mappings need which enrichments applied?

Thanks for the review :)
Mark

guegueng commented:
We haven't been formally keeping track of where enrichments are needed up to this point. I'll defer to the others who are programmers on the use of comments. 

I'm not sure if this documentation was shared with you, but it is a list of all of our current (or planned) enrichments. This was created for our partners, but it could be of use to you as you write mappings: https://digitalpubliclibraryofamerica.atlassian.net/wiki/pages/viewpage.action?pageId=27951124

anarchivist commented:
Also, regarding tracking enrichments, we have a couple existing tickets to address the management of enrichment profiles (see [#8173](https://issues.dp.la/issues/8173) and [#8164](https://issues.dp.la/issues/8164)).

marktriggs commented:
Thanks @guegueng and @anarchivist,
 
I've switched collection to use the right fields, and I've removed all the semicolon splitting and have put comments on each field that we'll need an enrichment for.  I guess we'll review these later when defining the enrichment profiles for each contributor.

markbreedlove commented:
I should be able to look at this early tomorrow ...

markbreedlove commented:
It looks like most of the questions have been answered by @anarchivist and @guegeng, but I hope this helps:

I think there should be separate tickets for the mappings and enrichments, in addition to the separate ticket that is for ingesting a given provider. So, right now, there's [Getty mapping (8120)](https://issues.dp.la/issues/8120) and [Ingest Getty (8119)](https://issues.dp.la/issues/8119), and I think a new ticket should be created for "Getty enrichment."

Meanwhile, it's always nice to have comments in the mappings that describe the original crosswalk specification for a field, and notes about how some field is not split, but will be during the enrichment.

markbreedlove commented:
@anarchivist would you like someone else to look at this? I'll just assign it back to you for routing.

anarchivist commented:
I'll kick this over to @no-reply since he'll be back in tomorrow.

no-reply commented on heidrun/getty.rb:
```diff
@@ -0,0 +1,127 @@
+  end
+
+  sourceResource class: DPLA::MAP::SourceResource do
+    # TODO: Enrichment needed to split title fields on semicolons
```
Should be set, here: https://github.com/dpla/KriKri/blob/develop/lib/krikri/enrichments/split_at_delimiter.rb

no-reply commented on heidrun/getty.rb:
```diff
@@ -0,0 +1,127 @@
+
+    dcformat record.field(*Krikri::PrimoParser.display('lds09'))
+
+    # TODO: Enrichment needed here to map (e.g.) 'Still image' to 'image'
```
We should be able to handle this with https://github.com/dpla/KriKri/blob/develop/lib/krikri/enrichments/dcmi_type_map.rb

no-reply commented on heidrun/mwdl.rb:
```diff
@@ -0,0 +1,124 @@
+    #
+    # Currently they are semicolon delimiter lists of years like:
+    #  1930; 1931; 1932; 1933; 1934; 1935; 1936; 1937; 1938; 1939; 1940; 1941;
+    #  1942; 1943; 1944; 1945; 1946"
```
I added a ticket for this here: https://issues.dp.la/issues/8186

no-reply commented on heidrun/mwdl.rb:
```diff
@@ -0,0 +1,124 @@
+      providedLabel subject
+    end
+
+    # TODO: Need an enrichment to parse these?  Currently may be semicolons, and
```
If I understand this comment correctly, the existing delimiter splitter and date parse tooling should handle this fine. Cases like `"20th Century"` may be somewhat buggy due to differences in the EDTF Century and Decade implementations, but I think it fails gracefully.

no-reply commented:
I made a pass through with comments. I think this looks good, except for a simplification at languages and a single enrichment we need to code up for MWDL (https://issues.dp.la/issues/8186).

Assigning this back to @marktriggs for revisions; once it's in, we should be ready to map and enrich Getty; MWDL will be dependant on the enrichment linked above.

marktriggs commented:
Great!  Thanks all.  I've just pushed the changes we needed.

I've left the enrichment TODO items in there for now, on the assumption that we can pull them out later once the enrichment profiles are done.  I thought it might be nice to have a checklist until then.

no-reply commented:
Looks good to me. :squash: :question:

marktriggs commented:
:elephant: (No pancake emoji, seriously?)
